### DISCUSSION TITLE
How to Omit Empty Members of a Struct When Serializing to JSON

### DISCUSSION THREAD

**Discussion by tomlin7** [nlohmann/json#2846] [Q&A] [answered]
Let me explain my case

I have a struct named `Person` as following
```cpp

struct Person
{
    std::string name;        // required
    int age;                 // optional

    NLOHMANN_DEFINE_TYPE_INTRUSIVE(Person, name, age)
};
```

So here in my case, the `age` is an optional field, so if a value is not passed for the `age`, then it should not be in the json, that is the result I'm expecting.

If I pass values for both fields of the struct, as following
```cpp
int main() {
    Person person{ "ned", 25 };

    json j;
    j = person;

    std::cout << j;

    return 0;
}
```
The result is 
```
{"age":25,"name":"ned"}
```

And if I pass only the required field (the `name`), as following
```cpp
int main() {
    Person person{ "ned" };

    json j;
    j = person;

    std::cout << j;

    return 0;
}
```
The result is 
```
{"age":0,"name":"ned"}
```
Here, the integer value defaulted to `0`.
What i want is to exclude the value from JSON if it's not passed at all.
Just like `omitempty` attributes, provided in other languages.

---

**Comment by nlohmann** [comment, marked as answer]
For this, I think the best way is to define the `from_json` and `to_json` functions yourself - the macros are just a convenient way to get some default behavior. See https://github.com/nlohmann/json#arbitrary-types-conversions.

**Reply by tomlin7** [reply]
Sure, I will define them my own in this case 👍🏻 Also I would like to know if something like `omitempty` is possible to implement, if so please try adding them too :)

**Reply by nlohmann** [reply]
I'll keep it in mind once I finally have a concept how to configure the dump function further.

**Reply by nlohmann** [reply]
In fact, you can implement an `omitempty` filter on to of the API. It would traverse your JSON value recursively and remove the unwanted values. The result can then be serialized as always.
